### PR TITLE
Add feature: Reopen closed tab/window in tab context menu.

### PR DESCRIPTION
**Pull request by XP1**
See "Restore previously closed tab: Context menu on tab #549".

### CODE REVIEW

**Review by gz83 — COMMENTED** [review]
Tabstrip is the main and commonly used interactive component. Any type of modification may cause exceptions. Please make sure to combine it with Thorium's code and conduct sufficient testing (you need to compile Thorium browser yourself) to avoid unnecessary trouble.

**Comment by gz83** [comment]
Is there any progress? If not, this PR will be closed.

**Comment by XP1** [comment]
The original commit worked on 120.0.6099.276, but 121 had a few changes, so I have merged and compiled with 121.0.6167.206 (d880ff5).

Will you try compiling also?

**Comment by Alex313031** [comment]
@XP1 @gz83 I am willing to accept this patch on the condition that you maintain it. It's already a lot of work for me to rebase thorium, so I don't wanna accept this if it means another 4 things I have to keep track of and rebase. How it would work is I would let you know when I start work on a new major revision, and let you know what that revision is, then you can update those 4 files and send in a new PR. I will compare the tab_strip_model.cc file (since I also modify it, and I don't want our stuff to conflict or override one another's patches) and then following testing and looking all good, I would merge it for each release.

What do you think?

**Review by gz83 — COMMENTED** [review]
Since the code review and testing have not yet been completed, I will temporarily vote -1 for this patch.

File: src/chrome/browser/ui/tabs/tab_menu_model.cc [hunk 1]
```diff
@@ -0,0 +1,243 @@
+// Copyright 2012 The Chromium Authors
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "chrome/browser/ui/tabs/tab_menu_model.h"
+
+#include "base/command_line.h"
+#include "base/feature_list.h"
+#include "base/i18n/rtl.h"
+#include "base/memory/raw_ptr.h"
```
Comment (gz83) [inline src/chrome/browser/ui/tabs/tab_menu_model.cc:10]: Add a header file on line 10

File: src/chrome/browser/ui/tabs/tab_menu_model.cc [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+#include "base/metrics/user_metrics.h"
+#include "build/build_config.h"
+#include "chrome/app/vector_icons/vector_icons.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/send_tab_to_self/send_tab_to_self_util.h"
+#include "chrome/browser/sessions/tab_restore_service_factory.h"
```
Comment (gz83) [inline src/chrome/browser/ui/tabs/tab_menu_model.cc:16]: Add a header file on line 16

File: src/chrome/browser/ui/tabs/tab_menu_model.cc [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+#include "chrome/browser/ui/tabs/tab_utils.h"
+#include "chrome/browser/ui/ui_features.h"
+#include "chrome/browser/ui/user_notes/user_notes_controller.h"
+#include "chrome/browser/ui/web_applications/web_app_tabbed_utils.h"
+#include "chrome/common/chrome_features.h"
+#include "chrome/grit/branded_strings.h"
+#include "chrome/grit/generated_resources.h"
+#include "components/feed/feed_feature_list.h"
+#include "components/reading_list/features/reading_list_switches.h"
+#include "components/sessions/core/tab_restore_service.h"
```
Comment (gz83) [inline src/chrome/browser/ui/tabs/tab_menu_model.cc:33]: Add a header file on line 33

File: src/chrome/browser/ui/tabs/tab_menu_model.cc [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    if (trs->entries().front()->type == sessions::TabRestoreService::WINDOW) {
+      id = IDS_REOPEN_WINDOW;
+    } else if (trs->entries().front()->type ==
+               sessions::TabRestoreService::GROUP) {
+      id = IDS_REOPEN_GROUP;
+    }
+  }
+
+  return id;
+}
```
Comment (gz83) [inline src/chrome/browser/ui/tabs/tab_menu_model.cc:72]: Line 55-line 72 add `TabMenuModel::GetRestoreTabCommandId` implementation

File: src/chrome/browser/ui/tabs/tab_menu_model.cc [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+  AddItemWithStringId(TabStripModel::CommandCloseTab, IDS_TAB_CXMENU_CLOSETAB);
+  AddItemWithStringId(TabStripModel::CommandCloseOtherTabs,
+                      IDS_TAB_CXMENU_CLOSEOTHERTABS);
+  AddItemWithStringId(TabStripModel::CommandCloseTabsToRight,
+                      base::i18n::IsRTL() ? IDS_TAB_CXMENU_CLOSETABSTOLEFT
+                                          : IDS_TAB_CXMENU_CLOSETABSTORIGHT);
+
+  AddSeparator(ui::NORMAL_SEPARATOR);
+  AddItemWithStringId(TabStripModel::CommandRestoreTab,
+                      GetRestoreTabCommandId(tab_strip->profile()));
```
Comment (gz83) [inline src/chrome/browser/ui/tabs/tab_menu_model.cc:239]: Line 236 - Line 238 calls the `TabMenuModel::GetRestoreTabCommandId` implementation

File: src/chrome/browser/ui/tabs/tab_menu_model.h [hunk 6]
```diff
@@ -0,0 +1,54 @@
+// Copyright 2011 The Chromium Authors
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef CHROME_BROWSER_UI_TABS_TAB_MENU_MODEL_H_
+#define CHROME_BROWSER_UI_TABS_TAB_MENU_MODEL_H_
+
+#include "base/memory/raw_ptr.h"
+#include "chrome/browser/profiles/profile.h"
```
Comment (gz83) [inline src/chrome/browser/ui/tabs/tab_menu_model.h:9]: Add a header file on line 9

File: src/chrome/browser/ui/tabs/tab_menu_model.h [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+               int index);
+  TabMenuModel(const TabMenuModel&) = delete;
+  TabMenuModel& operator=(const TabMenuModel&) = delete;
+  ~TabMenuModel() override;
+
+  DECLARE_CLASS_ELEMENT_IDENTIFIER_VALUE(kAddToNewGroupItemIdentifier);
+
+ private:
+  // Returns restore tab command id.
+  int GetRestoreTabCommandId(raw_ptr<Profile> profile) const;
```
Comment (gz83) [inline src/chrome/browser/ui/tabs/tab_menu_model.h:43]: Line 42 – Line 43 defines the `TabMenuModel::GetRestoreTabCommandId` private member

File: src/chrome/browser/ui/tabs/tab_strip_model.h [hunk 8]
```diff
@@ -0,0 +1,868 @@
...
+    CommandRemoveFromGroup,
+    CommandMoveToExistingWindow,
+    CommandMoveTabsToNewWindow,
+    CommandOrganizeTabs,
+    CommandFollowSite,
+    CommandUnfollowSite,
+    CommandCopyURL,
+    CommandGoBack,
+    CommandCloseAllTabs,
+    CommandRestoreTab,
```
Comment (gz83) [inline src/chrome/browser/ui/tabs/tab_strip_model.h:523]: Line 523 defines `CommandRestoreTab`

File: src/chrome/browser/ui/tabs/tab_strip_model.cc [hunk 9]
```diff
@@ -1701,6 +1721,9 @@ bool TabStripModel::ContextMenuCommandToBrowserCommand(int cmd_id,
     case CommandCloseTab:
       *browser_cmd = IDC_CLOSE_TAB;
       break;
+    case CommandRestoreTab:
+      *browser_cmd = IDC_RESTORE_TAB;
+      break;
```
Comment (gz83) [inline src/chrome/browser/ui/tabs/tab_strip_model.cc:1726]: LGTM

**Review by gz83 — COMMENTED** [review]
After testing, the newly added function works normally and can be adapted to multi-language environments. Everything is fine for me.

Since this patch involves new files, I have pointed out the code modifications involved in the new files. Alex, can you take a look?

In addition, if you need to submit some patches involving new functions in the future, you need to attach more intuitive test results, such as showing the test results in the form of a video.

@XP1 @Alex313031

**Comment by Alex313031** [comment]
@gz83 Why did you close it? And what do you mean he needs to add header files?

**Comment by gz83** [comment]
> @gz83 Why did you close it? And what do you mean he needs to add header files?

I didn't see a reply from him, so I closed it for now.  We can always turn it back on if needed.
In addition, the comments I made earlier were to point out to you the modifications he made (the newly added files, the specific modifications cannot be seen in GitHub)

@Alex313031

**Pull request closed without merging**